src/main/java/gregtech/api/recipes/ui/RecipeMapUI.java

```diff
@@ -0,0 +1,459 @@
+package gregtech.api.recipes.ui;
+
+import gregtech.api.capability.impl.FluidTankList;
+import gregtech.api.gui.GuiTextures;
+import gregtech.api.gui.ModularUI;
+import gregtech.api.gui.resources.TextureArea;
+import gregtech.api.gui.widgets.ProgressWidget;
+import gregtech.api.gui.widgets.RecipeProgressWidget;
+import gregtech.api.gui.widgets.SlotWidget;
+import gregtech.api.gui.widgets.TankWidget;
+import gregtech.api.recipes.Recipe;
+import gregtech.api.recipes.RecipeMap;
+
+import net.minecraftforge.items.IItemHandlerModifiable;
+
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectOpenHashMap;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.DoubleSupplier;
+
+public class RecipeMapUI<R extends RecipeMap<?>> {
```

Comment:
Should put an `@ApiStatus.Experimental` since this is going to change again soon